Make NTP client more robust

sylane opened the pull request:
The NTP client is now a state machine, use exponencial backoff, reuse the same IP check function than the client and periodically refresh the time.

One of the question is if refreshing the time every 256 seconds is good, a real NTP client would do some smart calculation instead of just setting the time to the new one...

ziopio commented on src/grisp_connect.app.src:
```diff
@@ -21,6 +21,7 @@
         {port, 443},
         {connect, true}, % keeps a constant connection with grisp.io
         {ntp, false}, % if set to true, starts the NTP client
+        {ntp_servers, ["0.europe.pool.ntp.org"]},
```
We document all `grisp_connect` options in the Readme file, make sure this and others are up to date.

GwendalLaurent commented on src/grisp_connect_ntp.erl:
```diff
@@ -1,22 +1,45 @@
+
+-define(NTP_PORT, 123). % NTP's UDP port
+-define(EPOCH, 2208988800). % offset yr 1900 to unix epochù
+-define(REFRESH_TIMEOUT, 1000 * 256). % ms
```
That's not needed for this PR, but we should allow the user to configure the REFRESH_TIMEOUT at some point. For example on the grisp_charger we will have a RTC and we won't need to refresh the internal time that often.According to [Wikipedia](https://en.wikipedia.org/wiki/Real-time_clock#Timing) a typical RTC has a drift of 1,7-8,6 second per day

GwendalLaurent commented on src/grisp_connect_ntp.erl:
```diff
@@ -27,82 +50,127 @@ get_time() ->
     end;
+?HANDLE_COMMON.
+
+refresh_time(enter, _OldState, #data{retry_count = RetryCount}) ->
+    Delay = grisp_connect_utils:retry_delay(RetryCount),
+    {keep_state_and_data, [{state_timeout, Delay, request_time}]};
+refresh_time(state_timeout, request_time,
+             Data = #data{retry_count = RetryCount}) ->
+    NTPServer = random_ntp_server(),
+    case refresh_current_time(NTPServer) of
+        {error, Reason} ->
+            ?LOG_INFO("Failed to get time from NTP server ~s: ~w",
+                      [NTPServer, Reason]),
+            {next_state, waiting_ip, Data#data{retry_count = RetryCount + 1}};
+        {ok, Datetime} ->
+            ?LOG_NOTICE("GRiSP clock set from NTP to ~s",
+                        [format_datetime(Datetime)]),
+            {next_state, ready, Data#data{retry_count = 0}}
+    end.
```
What happens if we get a `{get_time, Host}` call in this state ? I think we are missing a `?HANDLE_COMMON` at the end of the state function

sylane replied:
Very good catch !

GwendalLaurent commented on src/grisp_connect_ntp.erl:
```diff
@@ -27,82 +50,127 @@ get_time() ->
+    ?LOG_DEBUG("Schedule NTP time refresh in ~w ms", [?REFRESH_TIMEOUT]),
+    {keep_state_and_data, [{state_timeout, ?REFRESH_TIMEOUT, refresh_time}]};
+ready({call, From}, {get_time, Host}, _Data) ->
+    Reply = case do_get_time(Host) of
```
Should we also refresh the time of the GRiSP if we get this call ? Since we fetch the time using NTP we could as well refresh the time with the latest value we get using NTP

sylane replied:
I keep it as it was before, only as a helper/debug function to perform an NTP request...

sylane replied:
I wan't sure about changing the behavior of the call...

sylane commented:
> @sylane I think just a quick explanation of why you switched to a state machine and other things would be nice. Just to get a grasp on the intention of the changes here.

@GalaxyGorilla This refactor intention is to make the NTP client more robust. We enabled it in grisp_demo and we had some random crashes. It was already a state machine I just changed it to use the State function model so the code is more clear. I added exponential backoff too, added refresh and some parametrisation.
